### ISSUE TITLE
New lint: suggest `ptr::read` instead of `mem::replace(..., uninitialized())`

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-clippy#5575]
I have now at least twice seen code like this:
```rust
mem::replace(x, mem::uninitialized());
```
Most recently [here](https://github.com/liebharc/basic_dsp/blob/df92d9237251db18d5ee8dfd25c298c287e0a099/matrix/src/lib.rs#L234), but I saw this before somewhere (not sure where).

Could Clippy suggest to replace this code by the following:
```rust
ptr::read(x)
```

The latter is more clear, avoids the mess that is uninitialized memory (both of the cases where I saw this were actually UB because of this), and can never be wrong when the former is right. Indeed, LLVM will likely (but not always) compile the former to the latter, because "write of uninitialized data to memory" can be just optimized away to nothing.

**Comment by esamudera** [comment]
Hello, I'm interested in working on this issue, if this is still available.

All I need to do should be just:
- Add `mem::replace(x, mem::uninitialized());` test case to `tests/ui/mem_replace.rs`
- Implement lint in `clippy_lints/src/mem_replace.rs`: 
  - detect `mem::replace` fn call with `mem::unititialized()` param
  - suggest replacing it with `ptr::read(x)`

Right? If yes, I can start working on this :+1:

**Comment by RalfJung** [comment]
Yes that sounds good. :-)
For bonus points, also detect `MaybeUninit::uninit().assume_init()` as an equivalent to `mem::uninitialized()`.

**Comment by esamudera** [comment]
Okay, adding that to the test case :+1:
